### PR TITLE
is_placeholder() implemented

### PR DESCRIPTION
**Pull request by ahundt**
small part of #6928 and #7046 

### [is_placeholder()](https://github.com/fchollet/keras/pull/6928/files#diff-c66ff60af92d5eee116b055bbb711174R111)

I use is_placeholder() to [differentiate numpy inputs from tensorflow op inputs](https://github.com/fchollet/keras/pull/6928/files#diff-fe3ca734e8421a17729d0b0ea4c800baR1624), which allows [`y_true` and `y_pred` to be correctly assigned to the appropriate tf op](https://github.com/fchollet/keras/pull/6928/files#diff-299cfd5886683a4b012f286403769fc1R915), without this change the program crashes by trying to [incorrectly pass an op to K.placeholder](https://github.com/fchollet/keras/pull/6928/files#diff-299cfd5886683a4b012f286403769fc1R896).

`is_placeholder` can also be useful when enumerating the layers, inputs, outputs, and transforming an arbitrary model. For example anything that walks the graph may or may not want to modify placeholders, such as an automated labeling model to segmentation model converter.

The implementation of `is_placeholder` was based on [_uses_learning_phase and _keras_shape](https://github.com/fchollet/keras/pull/6928/files#diff-ff634bb5c5441d7052449f20018872b8R223).

### CODE REVIEW

**Comment by hgaiser** [comment]
Can you explain the use of `is_placeholder` to me? I noticed its use in `examples/variational_autoencoder.py` and looked up where it is used in master, looked through the changes of your PR, but its still unclear what it does. Why is it necessary in that example and when should I use it?

File: keras/backend/cntk_backend.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 import cntk as C
 import numpy as np
 from .common import _FLOATX, _EPSILON, image_dim_ordering, image_data_format
+from .common import is_placeholder
```
Comment (PavlosMelissinos) [inline keras/backend/cntk_backend.py]: `from .common import is_placeholder`

is this necessary?

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -14,6 +14,7 @@
 from .common import floatx
 from .common import _EPSILON
 from .common import image_data_format
+from .common import is_placeholder
```
Comment (PavlosMelissinos) [inline keras/backend/tensorflow_backend.py]: `from .common import is_placeholder`

Again, is this necessary?

File: keras/backend/theano_backend.py [hunk 3]
```diff
@@ -17,6 +17,7 @@
 
 import numpy as np
 from .common import _FLOATX, floatx, _EPSILON, image_data_format
+from .common import is_placeholder
```
Comment (PavlosMelissinos) [inline keras/backend/theano_backend.py]: necessary?

**Comment by PavlosMelissinos** [comment]
is_placeholder = True makes sense for tensors but why would layers also have the option?

File: keras/backend/cntk_backend.py (hunk 1 above)
Comment (ahundt) [inline keras/backend/cntk_backend.py]: Yes, I think so, how else could it be provided in the backend? Is there a better option?

File: keras/backend/tensorflow_backend.py (hunk 2 above)
Comment (ahundt) [inline keras/backend/tensorflow_backend.py]: [will follow conclusion from linked conversation](https://github.com/fchollet/keras/pull/7066/files/6c776c760e3a591fe930a5db0f628d7cb0450484#r128470161)

File: keras/backend/theano_backend.py (hunk 3 above)
Comment (ahundt) [inline keras/backend/theano_backend.py]: [will follow conclusion from linked conversation](https://github.com/fchollet/keras/pull/7066/files/6c776c760e3a591fe930a5db0f628d7cb0450484#r128470161)

File: keras/backend/cntk_backend.py (hunk 1 above)
Comment (PavlosMelissinos) [inline keras/backend/cntk_backend.py]: So you add it there once and it's available for every backend.

Comment (ahundt) [inline keras/backend/cntk_backend.py]: good point. Removed.

File: keras/backend/tensorflow_backend.py (hunk 2 above)
Comment (ahundt) [inline keras/backend/tensorflow_backend.py]: resolved

File: keras/backend/theano_backend.py (hunk 3 above)
Comment (ahundt) [inline keras/backend/theano_backend.py]: resolved

**Comment by ahundt** [comment]
@PavlosMelissinos Could you confirm that I've addressed everything? Thanks for the review!

**Comment by PavlosMelissinos** [comment]
1. `_is_placeholder` is a Layer/Node attribute. Therefore K.is_placeholder accepts a tensor, a layer or a node.
a. I think the docstring should be updated to reflect this.
b. Since the `tensor` parameter can either be a tensor or a layer/node object, do you prefer it like this or do you think it should be a class method of Layer/Node? In the second case, the calls would become `layer.is_placeholder()`, so you'd get away with having to define the parameters and it would still work for any backend.
2. For which layers/nodes/tensors should `_is_placeholder` be set to `True`? It is False by default, so in case the user would like to enable it, the 'writing your own keras layers' page should be updated with a brief description of the role of placeholder.

**Comment by ahundt** [comment]
> I think the docstring and the parameter name should be updated to reflect this.

How about this?

```
Returns if a Keras Tensor, Layer, or Node contains a `K.placeholder()`.
```

> do you think it should be a class method of Layer/Node?

I think keeping `is_placeholder()` separate lowers the cognitive load for users, they don't have to know about it unless they need it and putting it in the Layer API might be overly prominent.

> For which layers/nodes/tensors should _is_placeholder be set to True?

It should be `True` when the entity has a `K.placeholder()` under the hood.

> 'writing your own keras layers' page should be updated with an example

Good idea, I'll add this when I have a chance.

**Comment by PavlosMelissinos** [comment]
> How about this?
> 
> Returns if a Keras Tensor, Layer, or Node contains a `K.placeholder()`.

Yeah, something like that or as a more complete keras docstring:

>    """is_placeholder function.
>    \# Arguments
>        x : Tensor, Layer or Node.
>    \# Returns
>        True if `x` is a Keras Tensor, Layer or Node and contains a `K.placeholder()`, False otherwise.
>    """

> I think keeping is_placeholder() separate lowers the cognitive load for users, they don't have to know about it unless they need it and putting it in the Layer API might be overly prominent.

I'm not sure that's the case:
1. cognitive load seems the same to me, except for code completion maybe where there'll be one extra entry.
2. most users won't have to know about it since it's going to be a member method of an ancestor class (Layer/Node/Tensor).
3. I don't know about prominence, it's just a few extra lines.

My main objection is redundancy because it would have to be implemented three times instead of once.

I think I'm slightly in favor of turning it into a class member mostly due to encapsulation but redundancy is an issue so yeah, it's not a big deal, either solution sounds ok to me!

> It should be True when the entity has a K.placeholder() under the hood.

Last (tangential and slightly off-topic) question: Would it make sense to determine this on the spot? Flags are fast (they don't require any computation) but they're more prone to bugs in general. I know the flag was there before the PR and this is probably out of scope but I'm just wondering if maybe there's a better way.

File: keras/engine/topology.py [hunk 4]
```diff
@@ -1621,7 +1635,7 @@ def __init__(self, inputs, outputs, name=None):
                                                    i,
                                                    layer.__class__.__name__))
             self.input_names.append(layer.name)
-            if layer.is_placeholder:
+            if K.is_placeholder(layer):
```
Comment (fchollet) [inline keras/engine/topology.py]: `if getattr(layer, 'is_placeholder', False):`

**Comment by ahundt** [comment]
Based on a conversation with @fchollet there is no longer an independent `_is_placeholder()` function, since it can be substituted easily with `getattr(keras_placeholder, 'is_placeholder', False)`. Therefore, the separate function has been removed, and all areas correcting where `object.is_placeholder` was not set have been merged to #7067. Closing this pull request accordingly.

**Pull request closed without merging**